### ISSUE TITLE
Support for banded matrix operations

### ISSUE THREAD

**Issue by al5250** [pytorch/pytorch#118225]
### 🚀 The feature, motivation and pitch

It would be amazing if there could be support for [banded matrix](https://en.wikipedia.org/wiki/Band_matrix) operations, which are more efficiently stored by their diagonals rather than full matrices.  Some common operations include [`solve_banded`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solve_banded.html#scipy.linalg.solve_banded) (solving banded linear systems), [`cholesky_banded`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.cholesky_banded.html#scipy.linalg.cholesky_banded) (cholesky decompostition of banded matrices) and [`eig_banded`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eig_banded.html#scipy.linalg.eig_banded) (eigenvalues of banded matrices).  

### Additional context

The aformentioned functions are all currently implemented in `scipy.linalg`, which I believe calls Lapack under the hood.  TensorFlow also has a [`banded_triangular_solve`](https://www.tensorflow.org/api_docs/python/tf/linalg/banded_triangular_solve) that can perform batch processing.  However, none of these have GPU support.

**Comment by lezcano** [comment]
What's the support for these in Magma? I reckon that cuBlas doesn't have these.

**Comment by mfauvel** [comment]
Dear @al5250 ,

I have a pure pytorch implementation (work in progress) here for symmetric banded matrix, if it can be of any help:

https://src.koda.cnrs.fr/mathieu.fauvel.4/hp-filter/-/blob/main/src/hp_filter/banded_matrix.py

It is not really faster than conventional torch solver, but it consumed much less memory. I use it for banded symmetric linear problems found in the Hodrick–Prescott filter. I have also the derivative of the linear problem implemented  using same tools.

https://src.koda.cnrs.fr/mathieu.fauvel.4/hp-filter/-/blob/main/src/hp_filter/banded_linear_solver.py

Best

**Comment by nikitaved** [comment]
Have you also looked at https://github.com/cornellius-gp/linear_operator?

**Comment by Balandat** [comment]
Yeah, something like this could fit neatly into `linear_operator`.  The main benefit from that would be that due to how we use `__torch_function__`, you could do something like this:

```
banded_matrix = BandedLinearOperator(...)
torch.linalg.solve(banded_matrix, rhs)
```

and `linear_operator` would automatically dispatch to the underlying efficient banded solver.

**Comment by mfauvel** [comment]
Thanks for the update, I was aware of `linear_operator` : I had discussed some issues last year (i did not found anymore my issue, maybe it was on another github ?).

I am currently trying to fasten cholesky decomposition and solver with KeOps. If it is successful, I would consider to propose an extension, I like very much `linear_operator` ! In the mean time, if you do it better than me, please tell me :). Currently, reducing the memory overhead was my main motivation.

Yet, my work concerns only Symmetric Banded Matrix, not general Banded Matrix.
